projects/npm-tools/packages/npm-scripts/src/utils/getMergedConfig.js

```diff
@@ -153,17 +160,31 @@ function getMergedConfig(type, property) {
 			break;
 
 		case 'npmscripts': {
-			let presetConfig = {};
+			const rootDir = findRoot();
 
-			let userConfig = getUserConfig('npmscripts');
+			let rootDefaults;
 
-			// Use default config if no user config exists
+			try {
+				rootDefaults = require(path.join(
+					rootDir,
+					'npmscripts.config.js'
+				)).defaults;
+			} catch (err) {
+				if (err.code !== 'MODULE_NOT_FOUND') {
+					throw err;
+				}
 
-			if (Object.keys(userConfig).length === 0) {
-				userConfig = require('../config/npmscripts.config');
+				rootDefaults = {};
 			}
 
+			let userConfig = deepMerge([
+				require('../config/npmscripts.config.js'),
+				rootDefaults,
+				getUserConfig('npmscripts'),
+			]);
```

Comment:
I am not a huge fan of the override model here and it would be good to be really _sure_ that this is the one we want before locking it in. I haven't been able to think about this as deeply as you, so bear with me because there might be some mistakes or inaccuracies in what I am about to say.

I think I get _why_ you did it this way: ie. because there are 23 `npmscripts.config.js` files in liferay-portal (including the root one), and you don't want to do a blindly squish them together with the root because the overrides may produce unpredictable results. So, instead of that, you do this special merge that treats the `defaults` key from the top-level as special, and you carefully merge that in with the user config.

But the counter-argument against this design is that it is atypical (compared to other JS tools that do  config file inheritance, most canonically represented by ESLint's approach) and therefore quite subtle. As in, instead of simple/generic merging strategy, we have a "complex"/specific one. 😢

When I inspect the existing `npmscripts.config.js`, I don't see anything in there that looks unsafe to merge. These two are about as complicated as it gets (I'm just copy-pasting these from `master`, not from your other in-flight PR):

`modules/apps/document-library/document-library-web/npmscripts.config.js`:

```js
const preset = require('@liferay/npm-scripts/src/presets/standard');

module.exports = {
	build: {
		dependencies: [...preset.build.dependencies, 'asset-taglib'],
	},
	preset: '@liferay/npm-scripts/src/presets/standard',
};
```

`modules/apps/asset/asset-categories-admin-web/npmscripts.config.js`:

```js
module.exports = {
	federation: {
		exposes: [
			'<inputDir>/js/ActionsComponentPropsTransformer.js',
			'<inputDir>/js/AssetCategoriesManagementToolbarDefaultEventHandler.es.js',
			'<inputDir>/js/AssetCategoriesSelectorTag.es.js',
		],
		remotes: ['asset-taglib'],
		runLegacyBuild: false,
	},
	preset: '@liferay/npm-scripts/src/presets/standard',
};
```

We would just need to define the semantics to be (_I think_) "more local config overwrites more general config", which is something that `deepMerge` can do (ie. `deepMerge.MODE.OVERWRITE_ARRAYS`, and if that isn't exactly what we need, we can easily enough make another). Your [other PR](https://github.com/liferay-frontend/liferay-portal/blob/ee7d34a8190300fe146d89a5355578fe4eca0d49/modules/npmscripts.config.js) only defines `mode`, `remotes` and `shared` keys under `federation`. As in, a more local config would safely overwrite `mode` and I gather that we just need to decide whether arrays should be overwritten or appended (because `remotes` seems to be the common key that is going to come up). I'm sure you've analyzed this in detail and can know exactly what my clash and why it would be a problem if we didn't merge it right.

Reply:
Well, the reason was simpler than that...

We have definitions in the root `npmscripts.config.js` to do specific "global" tasks:

```javascript
const CHECK_AND_FIX_GLOBS = [
     '!settings.json',
     '!tsconfig.json',
     '/{,dxp/}*.{js,ts}',
     '/{,dxp/}apps/*/*/*.{js,json,ts,tsx}',
     '/{,dxp/}apps/*/*/*/*.{js,json,ts,tsx}',
     '/{,dxp/}apps/*/*/*/{src,test}/**/*.{js,scss,ts,tsx}',
     '/{,dxp/}apps/*/*/*/{src}/**/*.{jsp,jspf}',
     '/{,dxp/}apps/*/*/{src,test}/**/*.{js,scss,ts,tsx}',
     '/{,dxp/}apps/*/*/{src}/**/*.{jsp,jspf}',
];
```

In particular, these globs :point_up: are broader than the ones used in the default preset. Also, because we were already using that file, I didn't want to change that. 

But as you say, maybe this is the right time to do that change. Now that the file hasn't grown too big and we can still do it. I wasn't a big fan of the `defaults` field either for the same reasons you give.

The only caveat is, what would be the behavior of the root level (`liferay-portal/modules`) npm scripts... 

Currently we have:

```json
     "scripts": {
         "checkFormat": "liferay-npm-scripts check",
         "format": "liferay-npm-scripts fix"
     }
```

If we put defaults in the sibling `npmscripts.config.js` file it looks a bit incorrect to put these scripts here, because, being at the root level, they are intended for global purposes and, in that case, they may need some way to override defaults (as is the case here). So, a bit of conflict here, I would say...

Reply:
Regarding the merging strategy I've come across that in my local WIP and it seems we won't be able to use `deepMerge.MODE.OVERWRITE_ARRAYS` because of the `remotes` and `shared` arrays, unless we want to replicate them entirely each time we want to add a dependency. You could argue that you may as well need to remove a dependency and then you cannot use an extend-only strategy, but for now that doesn't happen and I think we can live with that, because if you need to remove something from the defaults, why did you put it in the defaults in first place? :sweat_smile:

Reply:
A good way to fix the global lint configuration "problem" is to provide an CLI argument for npm-scripts to specify a different config file (like some other tools do). That way we may use the npmscripts.config.js file in portal/modules as base config and provide another file for the global tasks (if needed).

Reply:
So we could have `modules/npmscripts.config.js` and `modules/npmscripts.config2.js`?

(Just kidding about the names). One of the reasons you wanted to put this stuff into `npmscripts.config.js` in the first place was to avoid the need to add another configuration file... So I think you need to make a stronger argument for how this would work and why it would be a good idea (ie. details of what the file would be called, where it would live, how it would be clearly separated from the other `npmscripts.config.js` at a conceptual level etc.)

Reply:
The other alternative is to refactor the whole `check` and `fix` for the workspace folder so that they simply scan subfolders and invoke the task per project (with different configurations for each project). However, we can achieve the same result with gradle (in the portal).

Shall we do the refactor? That would be the best option IMO, because it is the most standard behavior. However it is out of the scope of this PR, I think. Maybe we want to defer it. I don't know...

Reply:
I have finally added a `global` section to invert the relation between defaults and configuration used to check and lint the whole workspace.